### PR TITLE
Add 16550A UART console support

### PR DESCRIPTION
**Pull request by marceg**
HTIF is deprecated, and tohost/fromhost locations are no longer fixed.
UART is standard and more easily supported with recent bootloaders.
(Linux can use either virtio or this UART.)

### CODE REVIEW

**Comment by gktrk** [comment]
Sorry for the delay, I didn't get a chance to look at this until today. I just applied the patch and can confirm that it boots just fine. I'll try to do a code review in the next few days.

**Comment by gktrk** [comment]
@marceg Did you have to patch the BBL to get this working? Normally we take the following patch by Fabrice Bellard:

```
diff --git a/Makefile.in b/Makefile.in
index aeabdc2..f6b8c2f 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -61,7 +61,7 @@ else
   march := -march=@WITH_ARCH@
 endif
 is_32bit := $(findstring 32,$(march))
-mabi := -mabi=$(if $(is_32bit),ilp32,lp64)
+mabi := -mabi=$(if $(is_32bit),ilp32d,lp64d)
 
 # Installation directories
 
diff --git a/bbl/bbl.lds b/bbl/bbl.lds
index 26f5816..615c3dc 100644
--- a/bbl/bbl.lds
+++ b/bbl/bbl.lds
@@ -43,15 +43,10 @@ SECTIONS
   _etext = .;
 
   /*--------------------------------------------------------------------*/
-  /* HTIF, isolated onto separate page                                  */
+  /* HTIF I/Os                                                          */
   /*--------------------------------------------------------------------*/
-  . = ALIGN(0x1000);
-  .htif :
-  {
-    PROVIDE( __htif_base = .);
-    *(.htif)
-  }
-  . = ALIGN(0x1000);
+  tohost = 0x40008000;
+  fromhost = 0x40008008;
 
   /*--------------------------------------------------------------------*/
   /* Initialized data segment                                           */
diff --git a/machine/fdt.c b/machine/fdt.c
index e8a504f..260fae8 100644
--- a/machine/fdt.c
+++ b/machine/fdt.c
@@ -646,8 +646,12 @@ static bool hart_filter_mask(const struct hart_filter *filter)
 {
   if (filter->mmu_type == NULL) return true;
   if (strcmp(filter->status, "okay")) return true;
+#if __riscv_xlen == 32
+  if (!strcmp(filter->mmu_type, "riscv,sv32")) return false;
+#else
   if (!strcmp(filter->mmu_type, "riscv,sv39")) return false;
   if (!strcmp(filter->mmu_type, "riscv,sv48")) return false;
+#endif
   printm("hart_filter_mask saw unknown hart type: status=\"%s\", mmu_type=\"%s\"\n",
          filter->status, filter->mmu_type);
   return true;
diff --git a/machine/htif.c b/machine/htif.c
index aae8c56..360cbb4 100644
--- a/machine/htif.c
+++ b/machine/htif.c
@@ -7,19 +7,12 @@
 #include "syscall.h"
 #include <string.h>
 
-extern uint64_t __htif_base;
-volatile uint64_t tohost __attribute__((section(".htif")));
-volatile uint64_t fromhost __attribute__((section(".htif")));
+volatile extern uint64_t tohost;
+volatile extern uint64_t fromhost;
 volatile int htif_console_buf;
 static spinlock_t htif_lock = SPINLOCK_INIT;
 uintptr_t htif;
 
-#define TOHOST(base_int)	(uint64_t *)(base_int + TOHOST_OFFSET)
-#define FROMHOST(base_int)	(uint64_t *)(base_int + FROMHOST_OFFSET)
-
-#define TOHOST_OFFSET		((uintptr_t)tohost - (uintptr_t)__htif_base)
-#define FROMHOST_OFFSET		((uintptr_t)fromhost - (uintptr_t)__htif_base)
-
 static void __check_fromhost()
 {
   uint64_t fh = fromhost;
@@ -91,7 +84,7 @@ void htif_syscall(uintptr_t arg)
 
 void htif_console_putchar(uint8_t ch)
 {
-#if __riscv_xlen == 32
+#if __riscv_xlen == 32 && 0
   // HTIF devices are not supported on RV32, so proxy a write system call
   volatile uint64_t magic_mem[8];
   magic_mem[0] = SYS_write;
diff --git a/machine/htif.h b/machine/htif.h
index a4e5e8e..e06440a 100644
--- a/machine/htif.h
+++ b/machine/htif.h
@@ -5,14 +5,9 @@
 
 #include <stdint.h>
 
-#if __riscv_xlen == 64
 # define TOHOST_CMD(dev, cmd, payload) \
   (((uint64_t)(dev) << 56) | ((uint64_t)(cmd) << 48) | (uint64_t)(payload))
-#else
-# define TOHOST_CMD(dev, cmd, payload) ({ \
-  if ((dev) || (cmd)) __builtin_trap(); \
-  (payload); })
-#endif
+
 #define FROMHOST_DEV(fromhost_value) ((uint64_t)(fromhost_value) >> 56)
 #define FROMHOST_CMD(fromhost_value) ((uint64_t)(fromhost_value) << 8 >> 56)
 #define FROMHOST_DATA(fromhost_value) ((uint64_t)(fromhost_value) << 16 >> 16)
diff --git a/machine/mentry.S b/machine/mentry.S
index 99e1b72..1aae2c1 100644
--- a/machine/mentry.S
+++ b/machine/mentry.S
@@ -220,6 +220,46 @@ __trap_from_machine_mode:
   jal trap_from_machine_mode
   j restore_regs
 
+#define HTIF_TOHOST   0x40008000
+
+putstring:
+  li a2, HTIF_TOHOST
+  li a3, (1 << 24) | (1 << 16)
+1:      
+  lbu a1, (a0)
+  beqz a1, 2f
+/* Note: we use 32 bit accesses to work in all base isa modes */
+  sw a1, 0(a2)
+  sw a3, 4(a2)
+/* wait until the char is handled */
+3:      
+  lw a1, 0(a2)
+  lw a4, 4(a2)
+  or a1, a1, a4
+  bnez a1, 3b
+  addi a0, a0, 1
+  j 1b
+2:
+  ret
+fatal_error:
+  jal putstring
+  # exit the emulator if possible
+  li a2, HTIF_TOHOST
+  li a1, 1
+  sw a1, 0(a2)
+  sw zero, 4(a2)
+1:  
+  wfi
+  j 1b
+
+rv32_expected_str:
+  .asciz "Must run this program on a 32 bit RISCV CPU - aborting\n"
+rv64_expected_str:
+  .asciz "Must run this program on a 64 bit RISCV CPU - aborting\n"
+rv128_detected_str:
+  .asciz "128 bit RISCV CPU detected: setting XLEN to 64 bits\n"
+  .align 4
+  
 do_reset:
   li x1, 0
   li x2, 0
@@ -255,6 +295,41 @@ do_reset:
   li x31, 0
   csrw mscratch, x0
 
+  # check MXL value and abort if it is not what we expect
+  mv s0, a0
+  mv s1, a1
+  csrr a0, misa
+  bgez a0, misa_rv32
+#if __riscv_xlen == 32
+  la a0, rv32_expected_str
+  j fatal_error
+#else
+  slli a0, a0, 1
+  bgez a0, misa_ok
+  # we are on a 128 bit CPU: we force MXL to 64 bit as a test
+  la a0, rv128_detected_str
+  jal putstring
+  li a0, 1
+  li a1, 126
+  # cannot use slli because no 128 bit support in gas
+  sll a0, a0, a1 
+  csrc misa, a0
+  # also set MSTATUS.SXL and MSTATUS.UXL to 64 bits
+  li a0, (3 << 32) | (3 << 34)
+  csrs mstatus, a0
+  li a0, (1 << 32) | (1 << 34)
+  csrc mstatus, a0
+  j misa_ok
+#endif    
+misa_rv32:
+#if __riscv_xlen == 64
+  la a0, rv64_expected_str
+  j fatal_error
+#endif
+misa_ok:
+  mv a0, s0
+  mv a1, s1
+
   # write mtvec and make sure it sticks
   la t0, trap_vector
   csrw mtvec, t0
diff --git a/machine/minit.c b/machine/minit.c
index 953e1ba..fb6efbe 100644
--- a/machine/minit.c
+++ b/machine/minit.c
@@ -156,7 +156,7 @@ void init_first_hart(uintptr_t hartid, uintptr_t dtb)
   query_uart(dtb);
   query_uart16550(dtb);
   query_htif(dtb);
-  printm("bbl loader\r\n");
+  //  printm("bbl loader\r\n");
 
   hart_init();
   hls_init(0); // this might get called again from parse_config_string
diff --git a/machine/mtrap.c b/machine/mtrap.c
index e48d3d5..fd78793 100644
--- a/machine/mtrap.c
+++ b/machine/mtrap.c
@@ -35,8 +35,13 @@ static uintptr_t mcall_console_putchar(uint8_t ch)
 
 void putstring(const char* s)
 {
-  while (*s)
-    mcall_console_putchar(*s++);
+    int c;
+    while (*s) {
+        c = *s++;
+        if (c == '\n')
+            mcall_console_putchar('\r');
+        mcall_console_putchar(c);
+    }
 }
 
 void vprintm(const char* s, va_list vl)
diff --git a/pk/pk.lds b/pk/pk.lds
index 52a7239..8fb2e68 100644
--- a/pk/pk.lds
+++ b/pk/pk.lds
@@ -43,15 +43,10 @@ SECTIONS
   _etext = .;
 
   /*--------------------------------------------------------------------*/
-  /* HTIF, isolated onto separate page                                  */
+  /* HTIF I/Os                                                          */
   /*--------------------------------------------------------------------*/
-  . = ALIGN(0x1000);
-  .htif :
-  {
-    PROVIDE( __htif_base = . );
-    *(.htif)
-  }
-  . = ALIGN(0x1000);
+  tohost = 0x40008000;
+  fromhost = 0x40008008;
 
   /*--------------------------------------------------------------------*/
   /* Initialized data segment                                           */
```

which hardcodes the `tohost` and `fromhost` addresses for HTIF in BBL.

**Comment by gktrk** [comment]
At the bare minimum, I still had to apply the following patch to BBL:

```
diff --git a/Makefile.in b/Makefile.in
index aeabdc2..f6b8c2f 100644
diff --git a/machine/fdt.c b/machine/fdt.c
index e8a504f..260fae8 100644
--- a/machine/fdt.c
+++ b/machine/fdt.c
@@ -646,8 +646,12 @@ static bool hart_filter_mask(const struct hart_filter *filter)
 {
   if (filter->mmu_type == NULL) return true;
   if (strcmp(filter->status, "okay")) return true;
+#if __riscv_xlen == 32
+  if (!strcmp(filter->mmu_type, "riscv,sv32")) return false;
+#else
   if (!strcmp(filter->mmu_type, "riscv,sv39")) return false;
   if (!strcmp(filter->mmu_type, "riscv,sv48")) return false;
+#endif
   printm("hart_filter_mask saw unknown hart type: status=\"%s\", mmu_type=\"%s\"\n",
          filter->status, filter->mmu_type);
   return true;
```

**Comment by gktrk** [comment]
> HTIF is deprecated

Can you cite a source for this? I see that coreboot [dropped HTIF support](https://review.coreboot.org/c/coreboot/+/15289) in 2017, claiming that it's deprecated. However, couple of months after that, QEMU [included](https://lists.gnu.org/archive/html/qemu-devel/2018-02/msg05789.html) HTIF support in their RISC-V port.

File: src/riscv_machine.c [hunk 1]
```diff
@@ -68,13 +84,18 @@ typedef struct RISCVMachine {
...
 #define VIRTIO_BASE_ADDR 0x40010000
 #define VIRTIO_SIZE      0x1000
-#define VIRTIO_IRQ       1
+#define VIRTIO_IRQ       2
 #define PLIC_BASE_ADDR 0x40100000
 #define PLIC_SIZE      0x00400000
 #define FRAMEBUFFER_BASE_ADDR 0x41000000
 
+
+/***************************   RTC   ***************************/
```
Comment (gktrk) [inline src/riscv_machine.c:97]: Can you split section headers like this that are not related to the UART implementation onto separate commit? One extra commit that says something like "Insert comments to mark the device specific code section" would do.

File: src/riscv_machine.c [hunk 2]
```diff
@@ -99,6 +120,190 @@ static uint64_t rtc_get_time(RISCVMachine *m)
...
+    }
+    return val;
+}
+
+static void console_write_char(RISCVMachine *s, uint8_t c)
+{
+    uint8_t buf[1];
+
+    buf[0] = c;
+    s->common.console->write_data(s->common.console->opaque, buf, 1);
```
Comment (gktrk) [inline src/riscv_machine.c:218]: Any particular reason why this hasn't been simplified like the following:

```suggestion
    s->common.console->write_data(s->common.console->opaque, &c, 1);
```

If it's meant to preserve the old code, since there is no `0xff` masking on `tohost`, it may as well go away.

File: src/riscv_machine.c [hunk 3]
```diff
@@ -99,6 +120,190 @@ static uint64_t rtc_get_time(RISCVMachine *m)
...
+
+    default:
+        val = 0;
+        break;
+
+    }
+    return val;
+}
+
+static void console_write_char(RISCVMachine *s, uint8_t c)
```
Comment (gktrk) [inline src/riscv_machine.c:213]: This might benefit from inlining:

```suggestion
static inline void console_write_char(RISCVMachine *s, uint8_t c)
```

File: src/riscv_machine.c [hunk 4]
```diff
@@ -99,6 +120,190 @@ static uint64_t rtc_get_time(RISCVMachine *m)
...
+{
+    uint8_t buf[1];
+
+    buf[0] = c;
+    s->common.console->write_data(s->common.console->opaque, buf, 1);
+}
+
+static void console_write_string(RISCVMachine *s, char *str)
```
Comment (gktrk) [inline src/riscv_machine.c:221]: The only location that calls this function is wrapped in `#if defined(UART_OUT_DEBUG)`. This function should also be wrapped in `#if defined(UART_OUT_DEBUG)` to indicate that it is only used by the debug functionality.

**Comment by marceg** [comment]
No, I made the fix so it would work with standard BBL.  I only tested 
RV64 however.

You might want to document this somewhere.  There's no way I could have 
known about this patch.  It's also really preferable to have something 
that works with standard software, so what about either having this 
patch or similar go permanently in BBL, or adjusting the simulator to 
work with the existing BBL?

M


Gokturk Yuksek wrote on 2019-09-23 17:03:

File: src/riscvemu.c [hunk 5]
```diff
@@ -583,12 +583,16 @@ void virt_machine_run(VirtMachine *m)
 #ifndef _WIN32
         if (m->console_dev && FD_ISSET(stdin_fd, &rfds)) {
             uint8_t buf[128];
-            int ret, len;
-            len = virtio_console_get_write_len(m->console_dev);
-            len = min_int(len, sizeof(buf));
+            int ret, len = sizeof(buf), ulen = uart_can_rx(m);
+	    if (virtio_console_can_write_data(m->console_dev))
```
Comment (gktrk) [inline src/riscvemu.c:587]: Please use 4 spaces instead of one tab character. This line is mixing tabs and spaces.

**Comment by marceg** [comment]
Perhaps that's because I only tested RV64.  BBL worked fine unchanged in 
that case.

M



Gokturk Yuksek wrote on 2019-09-23 17:06:

**Review by gktrk — CHANGES_REQUESTED** [review]
Thanks for the patch! I work quite a bit on riscv32, which requires a modified BBL. I was tired to having to make sure that the HTIF addresses are hardcoded properly in BBL every time I compile a custom one.

**Comment by gktrk** [comment]
>Perhaps that's because I only tested RV64.  BBL worked fine unchanged in
>that case.

That would make sense. The MMU patch that I mentioned is only required for riscv32.

**Comment by gktrk** [comment]
> You might want to document this somewhere.  There's no way I could have
> known about this patch.  It's also really preferable to have something
> that works with standard software, so what about either having this
> patch or similar go permanently in BBL, or adjusting the simulator to
> work with the existing BBL?

This was part of TinyEMU before we forked it. I'm not exactly sure why BBL doesn't support sv32. There's an issue created for it on github (https://github.com/riscv/riscv-pk/issues/160) but I see no activity on it.

After I merge your UART changes, the required BBL patch will substantially shrink. We will make sure to update the documentation regarding that for riscv32 systems.

**Comment by marceg** [comment]
> > HTIF is deprecated
> 
> Can you cite a source for this?

The fact that this simulator (nor the original TinyEMU from which it was derived) work with standard BBL was itself a hint.  (The patch you mentioned was nowhere in simulator or PK distros.  And no info on bellard.org about how anything used by TinyEMU was built, such as kernel, BBL, etc etc -- all essential info for anyone wanting to use this -- did it all from scratch.)

BBL itself has a comment in machine/htif.c:   // HTIF devices are not supported on RV32

I don't quite remember where I saw other references.  Quick googling for "HTIF deprecated" turns up things like this:   https://forums.sifive.com/t/failure-of-riscv-linux/170/3

In any case, given things that ought to work out-of-the-box did not work in several cases, that seemed to confirm whatever it is I had seen about it being deprecated.  Until there's a standard mechanism in BBL that allows it to support HTIF with any emulator/simulator, without having to patch it, I'm still interpreting the current situation as HTIF being deprecated.

Oh, and there's no HTIF support at all in OpenSBI.  AFAICT if you want any kind of standard mechanism, HTIF isn't enough.

HTH

**Comment by marceg** [comment]
> This was part of TinyEMU before we forked it. I'm not exactly sure why BBL doesn't support sv32. There's an issue created for it on github ([riscv/riscv-pk#160](https://github.com/riscv/riscv-pk/issues/160)) but I see no activity on it.

Strange.  Perhaps a pull request, explaining how this fixes SV32, will help?...

**Comment by gktrk** [comment]
Committed in https://github.com/bucaps/marss-riscv/commit/bc1479118f258fc9857480356d8edec23725890d. Thanks!

**Comment by gktrk** [comment]
> > > HTIF is deprecated
> > 
> > 
> > Can you cite a source for this?
> 
> The fact that this simulator (nor the original TinyEMU from which it was derived) work with standard BBL was itself a hint. (The patch you mentioned was nowhere in simulator or PK distros. And no info on bellard.org about how anything used by TinyEMU was built, such as kernel, BBL, etc etc -- all essential info for anyone wanting to use this -- did it all from scratch.)
> 

For the patch, you'd have to download Bellard's disk image: https://bellard.org/tinyemu/diskimage-linux-riscv-2018-09-23.tar.gz. Inside, there is a `patches` directory that contains a patch called `riscv-pk.diff`. It's really obscure.

> BBL itself has a comment in machine/htif.c: // HTIF devices are not supported on RV32
> 
> I don't quite remember where I saw other references. Quick googling for "HTIF deprecated" turns up things like this: https://forums.sifive.com/t/failure-of-riscv-linux/170/3
> 
> In any case, given things that ought to work out-of-the-box did not work in several cases, that seemed to confirm whatever it is I had seen about it being deprecated. Until there's a standard mechanism in BBL that allows it to support HTIF with any emulator/simulator, without having to patch it, I'm still interpreting the current situation as HTIF being deprecated.

I agree with the sentiment. Like I said before, I also had issues trying to get HTIF work on riscv32.

> 
> Oh, and there's no HTIF support at all in OpenSBI. AFAICT if you want any kind of standard mechanism, HTIF isn't enough.
> 
> HTH

**Comment by gktrk** [comment]
> > This was part of TinyEMU before we forked it. I'm not exactly sure why BBL doesn't support sv32. There's an issue created for it on github ([riscv/riscv-pk#160](https://github.com/riscv/riscv-pk/issues/160)) but I see no activity on it.
> 
> Strange. Perhaps a pull request, explaining how this fixes SV32, will help?...

See: https://github.com/bucaps/marss-riscv/issues/9
